proxy/src/metrics.rs

```diff
@@ -0,0 +1,200 @@
+            let mut value = *curr_val;
+
+            if let Some((prev_val, prev_time)) = cached_metrics.get(curr_key) {
+                // Only send metrics updates if the metric has changed
+                if curr_val - prev_val > 0 {
+                    value = curr_val - prev_val;
+                    start_time = *prev_time;
+                } else {
+                    return None;
+                }
+            };
+
+            Some(Event {
+                kind: EventType::Incremental {
+                    start_time,
+                    stop_time: *curr_time,
+                },
+                metric: PROXY_IO_BYTES_PER_CLIENT,
+                idempotency_key: idempotency_key(hostname.clone()),
+                value,
+                extra: Ids {
+                    endpoint_id: curr_key.endpoint_id.clone(),
+                    direction: curr_key.direction.clone(),
+                },
+            })
+        })
+        .collect();
+
+    if metrics_to_send.is_empty() {
+        trace!("no new metrics to send");
+        return Ok(());
+    }
+
+    // Send metrics.
+    // Split into chunks of 1000 metrics to avoid exceeding the max request size
+    for chunk in metrics_to_send.chunks(CHUNK_SIZE) {
+        let chunk_json = serde_json::value::to_raw_value(&EventChunk { events: chunk })
+            .expect("ProxyConsumptionMetric should not fail serialization");
+
+        let res = client
+            .post(metric_collection_endpoint.clone())
+            .json(&chunk_json)
+            .send()
+            .await;
+
+        match res {
+            Ok(res) => {
+                if res.status().is_success() {
+                    // update cached metrics after they were sent successfully
+                    for send_metric in chunk {
+                        let stop_time = match send_metric.kind {
+                            EventType::Incremental { stop_time, .. } => stop_time,
+                            _ => unreachable!(),
+                        };
+
+                        cached_metrics
+                            .entry(ProxyConsumptionMetricsKey {
+                                endpoint_id: send_metric.extra.endpoint_id.clone(),
+                                direction: send_metric.extra.direction.clone(),
+                            })
+                            // update cached value (add delta) and time
+                            .and_modify(|e| {
+                                e.0 += send_metric.value;
+                                e.1 = stop_time
+                            })
+                            // cache new metric
+                            .or_insert((send_metric.value, stop_time));
+                    }
+                } else {
+                    error!("metrics endpoint refused the sent metrics: {:?}", res);
+                }
+            }
+            Err(err) => {
+                error!("failed to send metrics: {:?}", err);
+            }
```

Comment:
It's possible to decrease nesting level here:

```rust
for ... {
    let Ok(res) = res else {
        error!("failed to send metrics: {:?}", err);
        continue;
    };
    
    // yay, res is available here!
}
```

Reply:
nice!
though, how do I get the `err` using this syntax?

Reply:
Ah, right... I'm getting tired.

```rust
let res = match res {
  Ok(x) => x,
  Err(e) => { error!(...); continue; }
}
```